source/common/network/dns_impl.h

```diff
@@ -34,28 +34,75 @@ class DnsResolverImpl : public DnsResolver, protected Logger::Loggable<Logger::I
   // Network::DnsResolver
   ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
                           ResolveCb callback) override;
+  ActiveDnsQuery* resolveSrv(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                             ResolveSrvCb callback) override;
 
 private:
   friend class DnsResolverImplPeer;
-  struct PendingResolution : public ActiveDnsQuery {
+  struct PendingResolutionBase : public ActiveDnsQuery {
     // Network::ActiveDnsQuery
-    PendingResolution(ResolveCb callback, Event::Dispatcher& dispatcher, ares_channel channel,
-                      const std::string& dns_name)
-        : callback_(callback), dispatcher_(dispatcher), channel_(channel), dns_name_(dns_name) {}
+    PendingResolutionBase(Event::Dispatcher& dispatcher, ares_channel channel,
+                          const std::string& dns_name)
+        : dispatcher_(dispatcher), channel_(channel), dns_name_(dns_name) {}
 
     void cancel() override {
       // c-ares only supports channel-wide cancellation, so we just allow the
       // network events to continue but don't invoke the callback on completion.
       cancelled_ = true;
     }
 
+    template <typename AddressInstance>
+    bool fireCallback(std::function<void(std::list<AddressInstance>&&)> callback,
```

Comment:
Prefer not to use templating and duck typing here if we could instead use inheritance.

Reply:
Addressed in ea167bc.